### PR TITLE
Fix: Reject all realization related to deleted activity or comment - MEED-1890 - Meeds-io/meeds#781

### PR DESCRIPTION
**Pull request by AzmiTouil**
This PR will ensure the rejection of  all achievements related to a deleted activity or comment

### CODE REVIEW

File: services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java [hunk 1]
```diff
@@ -194,6 +194,12 @@
             " AND g.receiver = :receiverId" +
             " AND g.objectId = :objectId"
 )
+
+@NamedQuery(
+    name = "GamificationActionsHistory.getRealizationsByObjectId",
+    query = "SELECT g FROM GamificationActionsHistory g" +
+            " WHERE g.objectId = :objectId"
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java]: We need the object type(s) here as well, isn't it ?

```suggestion
            " WHERE g.objectId = :objectId"
```

**Comment by exo-swf** [comment]
Your PR triggers too many exo-ci builds! Please finish your work and then, set your PR ready! Thank you

File: portlets/src/main/webapp/vue-app/realizations/components/RealizationItem.vue [hunk 2]
```diff
@@ -361,6 +371,23 @@ export default {
...
+      }
+      this.$nextTick(() => {
+        if (this.getLink) {
+          const initPromise = this.getLink(this.realization);
+          if (initPromise && initPromise.then) {
+            return initPromise
+              .then(() => this.$nextTick());
+          }
+        }
+      });
```
Comment (boubaker) [inline portlets/src/main/webapp/vue-app/realizations/components/RealizationItem.vue]:
```suggestion
      if (this.status === 'DELETED' || this.status === 'CANCELED') {
        this.$set(this.realization, 'link', null);
      } else if (!this.objectType && this.objectId) {
        this.$set(this.realization, 'link', this.objectId);
      } else if (this.getLink) {
        const linkPromise = this.getLink(this.realization);
        if (linkPromise?.then) {
          return linkPromise;
        }
      }
```

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java [hunk 3]
```diff
@@ -117,5 +117,16 @@ public interface GamificationConstant {
     String GAMIFICATION_DEFAULT_DATA_PREFIX                          = "def_";
 
 
+    //Achievement object types
+    
+    String IDENTITY_OBJECT_TYPE                                       = "identity";
+
+    String ACTIVITY_OBJECT_TYPE                                       = "activity";
+
+    //LOG event error
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]:
```suggestion
```

File: services/src/main/java/org/exoplatform/addons/gamification/service/configuration/RealizationsServiceImpl.java [hunk 4]
```diff
@@ -170,6 +171,7 @@ public GamificationActionsHistoryDTO updateRealization(GamificationActionsHistor
   }
 
   @Override
+  @ExoTransactional
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/addons/gamification/service/configuration/RealizationsServiceImpl.java]:
```suggestion
```
ExoTransactional has to be used on DAO layer instead if needed

File: portlets/src/main/webapp/vue-app/realizations/components/RealizationItem.vue (hunk 2 above)
Comment (AzmiTouil) [reply portlets/src/main/webapp/vue-app/realizations/components/RealizationItem.vue]: Done!

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java (hunk 3 above)
Comment (AzmiTouil) [reply services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]: Removed!

File: services/src/main/java/org/exoplatform/addons/gamification/service/configuration/RealizationsServiceImpl.java (hunk 4 above)
Comment (AzmiTouil) [reply services/src/main/java/org/exoplatform/addons/gamification/service/configuration/RealizationsServiceImpl.java]: Removed, thanks

File: services/src/main/java/org/exoplatform/addons/gamification/listener/social/space/GamificationSpaceListener.java [hunk 5]
```diff
@@ -144,30 +142,31 @@ public void spaceBannerEdited(SpaceLifeCycleEvent event) {
...
       gam.put("ruleTitle", ruleTitle);
-      gam.put("object", spaceURL);
+      gam.put("objectId", receiverId);
+      gam.put("objectType", IDENTITY_OBJECT_TYPE);
       gam.put("senderId", senderId);
       gam.put("receiverId", receiverId);
       listenerService.broadcast(GENERIC_EVENT_NAME, gam, null);
     } catch (Exception e) {
-      LOG.error("Cannot broadcast gamification event");
+      LOG.error(CANNOT_BROADCAST_GAMIFICATION_EVENT);
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/addons/gamification/listener/social/space/GamificationSpaceListener.java]: Please fix this old bug by avoiding exception swallow (same for other parts):

```suggestion
      LOG.error("Error while broadcasting gamification event: {}", gam, e);
```

**Comment by exo-swf** [comment]
Your PR triggers too many exo-ci builds! Please finish your work and then, set your PR ready! Thank you

**Review by boubaker — APPROVED** [review]

**Pull request merged**